src/util/ringdelaybuffer.cpp

```diff
@@ -0,0 +1,215 @@
+/*
+* $Id$
+* Portable Audio I/O Library
+* Ring Buffer utility.
+*
+* Author: Phil Burk, http://www.softsynth.com
+* modified for SMP safety on OS X by Bjorn Roche.
+* also allowed for const where possible.
+* modified for multiple-byte-sized data elements by Sven Fischer
+*
+* Note that this is safe only for a single-thread reader
+* and a single-thread writer.
+*
+* This program is distributed with the PortAudio Portable Audio Library.
+* For more information see: http://www.portaudio.com
+* Copyright (c) 1999-2000 Ross Bencina and Phil Burk
+*
+* Permission is hereby granted, free of charge, to any person obtaining
+* a copy of this software and associated documentation files
+* (the "Software"), to deal in the Software without restriction,
+* including without limitation the rights to use, copy, modify, merge,
+* publish, distribute, sublicense, and/or sell copies of the Software,
+* and to permit persons to whom the Software is furnished to do so,
+* subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be
+* included in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
+* ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+* CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+* WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+/*
+* The text above constitutes the entire PortAudio license; however,
+* the PortAudio community also makes the following non-binding requests:
+*
+* Any person wishing to distribute modifications to the Software is
+* requested to send the modifications to the original developer so that
+* they can be incorporated into the canonical version. It is also
+* requested that these non-binding requests be included along with the
+* license above.
+*/
+
+#include "ringdelaybuffer.h"
+
+RingDelayBuffer::RingDelayBuffer(SINT bufferSize)
+        : m_bufferSize(bufferSize),
+          m_readPos(0),
+          m_writePos(0),
+          m_ringFullMask((bufferSize * 2) - 1),
+          m_ringMask(bufferSize - 1),
+          m_jumpLeftAroundMask(0),
+          m_buffer(bufferSize) {
+    // TODO(davidchocholaty) Handle to allow only power of two for the size of the ring delay buffer.
+
+    // Set the ring delay buffer items to 0.
+    CSAMPLE* bufferData = m_buffer.data();
+    memset(bufferData, 0, sizeof(*bufferData) * bufferSize);
```

Comment:
This should be not necessary. The buffer should never return uninitialized zeros. In out of memory situation the usage code needs to fade to zero to avoid clicks.

Reply:
Initializing values on zero was created, due to it is possible to move with `m_readPos` to left over the left delay buffer bound. For the current code version, it is allowed in the beginning too. Depending on that, the read position can read values, which weren't written by `m_writePos`. If I'm thinking about it, not a good practice. It is possible to allow skipping to the left only if the items were written by `m_writePos` with `VERIFY_OR_DEBUG_ASSERT`.